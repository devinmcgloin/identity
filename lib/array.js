const flatten = (obj) => {
  let field = {};

  Object.entries(obj).forEach(([k, v]) => {
    typeof v === 'object' && v !== null && !Array.isArray(v)
      ? Object.assign(field, flatten(v))
      : Object.assign(field, { [k]: v });
  });
  return field;
};

var groupBy = (xs, key) => {
  return xs.reduce((rv, x) => {
    (rv[x[key]] = rv[x[key]] || []).push(x);
    return rv;
  }, {});
};

export { flatten, groupBy };
