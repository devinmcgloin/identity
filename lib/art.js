import walkDirectory from 'lib/walk';
import sizeOf from 'image-size';

const projectsDirectory = 'public/art';

export const getAllArtworkPaths = async () => {
  let images = walkDirectory(projectsDirectory);
  return images.map((fileName) => {
    let size = sizeOf(fileName);
    return { ...size, path: fileName.replace(/^public/, '') };
  });
};
