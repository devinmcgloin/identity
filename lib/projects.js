import fs from 'fs';
import path from 'path';
import matter from 'gray-matter';

const projectsDirectory = 'content/projects/';

export const getAllProjectIds = () => {
  const fileNames = fs.readdirSync(projectsDirectory);

  return fileNames.map((fileName) => {
    return {
      params: {
        slug: fileName.replace(/\.mdx$/, ''),
      },
    };
  });
};

export function getSortedProjectData() {
  const fileNames = fs.readdirSync(projectsDirectory);
  const allProjectData = fileNames.map((fileName) => {
    const id = fileName.replace(/\.mdx$/, '');

    const fullPath = path.join(projectsDirectory, fileName);
    const fileContents = fs.readFileSync(fullPath, 'utf8');

    const matterResult = matter(fileContents);
    const frontMatter = JSON.parse(JSON.stringify(matterResult));

    return {
      slug: `/projects/${id}`,
      id,
      ...frontMatter.data,
    };
  });

  return allProjectData.sort((a, b) => {
    if (a.date < b.date) {
      return 1;
    } else {
      return -1;
    }
  });
}
