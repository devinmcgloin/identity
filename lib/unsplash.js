import Unsplash, { toJson } from 'unsplash-js';

let unsplash;

const userStats = async () => {
  if (!unsplash) {
    unsplash = new Unsplash({
      accessKey: process.env.UNSPLASH_ACCESS_KEY,
    });
  }

  const userStats = await unsplash.users.statistics('devinmcgloin');
  const { downloads, views } = await toJson(userStats);

  let images = [];
  let currentBatchSize,
    batchSize = 10,
    page = 1;
  do {
    let currentBatch = await toJson(
      await unsplash.users.photos('devinmcgloin', page, batchSize, 'popular')
    );
    images = images.concat(currentBatch);
    currentBatchSize = currentBatch.length;
    page += 1;
  } while (batchSize == currentBatchSize);

  return {
    downloads: downloads.total,
    views: views.total,
    images: images,
  };
};

export { userStats };
