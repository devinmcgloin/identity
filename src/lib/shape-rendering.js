import { cordtopixel, p2c } from './math';

const renderCircle = (ctx, w, h, r, theta, s) => {
  var p = cordtopixel(w, h, p2c(r, theta));
  ctx.beginPath();
  ctx.arc(p[0], p[1], s, 0, Math.PI * 2, true);
  ctx.fill();
};

const BSpline = [
  -1 / 6,
  3 / 6,
  -3 / 6,
  1 / 6,
  3 / 6,
  -6 / 6,
  0,
  4 / 6,
  -3 / 6,
  3 / 6,
  3 / 6,
  1 / 6,
  1 / 6,
  0,
  0,
  0,
];

const cubic = (A, t) => {
  return A[0] * t * t * t + A[1] * t * t + A[2] * t + A[3];
};

const drawCurves = (ctx, w, h, C) => {
  var i,
    n,
    p,
    pix,
    x,
    y,
    z,
    fl = 5;

  // LOOP THROUGH CURVES.

  for (n = 0; n < C.length; n++) {
    // BUILD THE PROJECTED CURVE, POINT BY POINT.

    pix = [];
    for (i = 0; i < C[n].length; i++) {
      // TRANSFORM POINT

      x = C[n][i][0];
      y = C[n][i][1];
      z = C[n][i][2];
      if (z === undefined) z = 0;

      // DO PERSPECTIVE TRANSFORM

      x *= fl / (fl - z);
      y *= fl / (fl - z);

      // DO VIEWPORT TRANSFORM

      x = w * x * 0.5 + 0.5 * w;
      y = -w * y * 0.5 + 0.5 * h;
      pix.push([x, y]);
    }

    // DRAW THE PROJECTED CURVE ONTO THE CANVAS.
    for (i = 1; i < pix.length; i++) {
      ctx.beginPath();
      ctx.arc(pix[i][0], pix[i][1], 0.5, 0, Math.PI * 2);
      ctx.fill();
    }
  }
};

export { renderCircle, drawCurves, cubic, BSpline };
