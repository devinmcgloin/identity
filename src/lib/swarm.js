/* Core algorithm comes from https://github.com/skeeto/boids-js. Modified for my uses here */

// TODO set this up so that the number of boids in the swarm can be resized
// TODO allow boids to reproduce with new values

class Boid {
  constructor(swarm, type) {
    this.x = Math.random() * swarm.width;
    this.y = Math.random() * swarm.height;
    this.heading = Math.random() * 2 * Math.PI - Math.PI;
    this.type = type;
    this.fillStyle =
      type === 'salmon' ? '#fa9466' : type === 'pred' ? '#f44141' : '#f4bf42';
    this.radius = 6;
    this.speed = 2;
    this.radialSpeed = Math.PI / 60;
    this.vision = 50;
  }

  update = function(config) {
    this.fillStyle = config.fillStyle;
    this.radius = config.radius;
    this.speed = config.speed;
    this.radialSpeed = config.radialSpeed;
    this.vision = config.vision;
  };

  draw = function(ctx) {
    var pointLen = this.radius * 2.5;
    ctx.fillStyle = this.fillStyle;
    ctx.beginPath();
    ctx.arc(this.x, this.y, this.radius, 0, 2 * Math.PI);
    ctx.fill();
    ctx.beginPath();
    ctx.moveTo(
      this.x + Math.cos(this.heading + Math.PI / 2) * this.radius,
      this.y + Math.sin(this.heading + Math.PI / 2) * this.radius
    );
    ctx.lineTo(
      this.x + Math.cos(this.heading + Math.PI) * pointLen,
      this.y + Math.sin(this.heading + Math.PI) * pointLen
    );
    ctx.lineTo(
      this.x + Math.cos(this.heading - Math.PI / 2) * this.radius,
      this.y + Math.sin(this.heading - Math.PI / 2) * this.radius
    );
    ctx.fill();
  };

  distance = function(boid, width, height) {
    var x0 = Math.min(this.x, boid.x),
      x1 = Math.max(this.x, boid.x);
    var y0 = Math.min(this.y, boid.y),
      y1 = Math.max(this.y, boid.y);
    var dx = Math.min(x1 - x0, x0 + width - x1);
    var dy = Math.min(y1 - y0, y0 + height - y1);
    return Math.sqrt(dx * dx + dy * dy);
  };

  getNeighbors = function(swarm) {
    var w = swarm.width,
      h = swarm.height;
    var neighbors = [];

    for (var i = 0; i < swarm.boids.length; i++) {
      var boid = swarm.boids[i];

      if (this !== boid && this.distance(boid, w, h) < this.vision) {
        neighbors.push(boid);
      }
    }

    return neighbors;
  };

  static wrap = function(value) {
    var min, max;

    if (arguments.length === 2) {
      min = 0;
      max = arguments[1];
    } else if (arguments.length === 3) {
      min = arguments[1];
      max = arguments[2];
    } else {
      throw new Error('wrong number of arguments');
    }

    while (value >= max) value -= max - min;

    while (value < min) value += max - min;

    return value;
  };

  static clamp = function(value, limit) {
    return Math.min(limit, Math.max(-limit, value));
  };

  static meanAngle = function() {
    var sumx = 0,
      sumy = 0,
      len = arguments.length;

    for (var i = 0; i < len; i++) {
      sumx += Math.cos(arguments[i]);
      sumy += Math.sin(arguments[i]);
    }

    return Math.atan2(sumy / len, sumx / len);
  };

  step = function(swarm) {
    var w = swarm.width,
      h = swarm.height;
    var neighbors = this.getNeighbors(swarm);

    if (neighbors.length > 0) {
      var meanhx = 0,
        meanhy = 0;
      var meanx = 0,
        meany = 0;
      var mindist = this.radius * 2,
        min = null;

      for (var i = 0; i < neighbors.length; i++) {
        var boid = neighbors[i];
        meanhx += Math.cos(boid.heading);
        meanhy += Math.sin(boid.heading);
        meanx += boid.x;
        meany += boid.y;
        var dist = this.distance(boid, w, h);

        if (dist < mindist) {
          mindist = dist;
          min = boid;
        }
      }

      meanhx /= neighbors.length;
      meanhy /= neighbors.length;
      meanx /= neighbors.length;
      meany /= neighbors.length;
      var target;

      var nearPred =
        swarm.activePred &&
        this.type !== 'pred' &&
        this.distance(swarm.pred, w, h) < this.vision * 2;
      var eaten =
        swarm.activePred &&
        this.type !== 'pred' &&
        this.distance(swarm.pred, w, h) < swarm.pred.radius;
      if (min && this.type !== 'pred') {
        // Keep away!
        target = Math.atan2(this.y - min.y, this.x - min.x);
      } else {
        // Match heading and move towards center
        var meanh = Math.atan2(meanhy, meanhx);
        var center = Math.atan2(meany - this.y, meanx - this.x);
        target = Boid.meanAngle(meanh, meanh, meanh, center);
      }

      var delta;
      // Move in this direction
      if (eaten) {
        this.x = Math.random() * swarm.width;
        this.y = Math.random() * swarm.height;
      } else if (nearPred) {
        target = Math.atan2(this.y - swarm.pred.y, this.x - swarm.pred.x);
        delta = Boid.wrap(target - this.heading, -Math.PI, Math.PI);
        delta = Boid.clamp(delta, this.radialSpeed);
        this.heading = Boid.wrap(this.heading + delta, -Math.PI, Math.PI);
      } else {
        delta = Boid.wrap(target - this.heading, -Math.PI, Math.PI);
        delta = Boid.clamp(delta, this.radialSpeed);
        this.heading = Boid.wrap(this.heading + delta, -Math.PI, Math.PI);
      }
    }

    this.move(swarm);
  };

  move = function(swarm) {
    var padding = swarm.padding;
    var width = swarm.width,
      height = swarm.height;
    this.x = Boid.wrap(
      this.x + Math.cos(this.heading) * this.speed,
      -padding,
      width + padding * 2
    );
    this.y = Boid.wrap(
      this.y + Math.sin(this.heading) * this.speed,
      -padding,
      height + padding * 2
    );
  };
}

class Swarm {
  constructor(ctx) {
    this.ctx = ctx;
    this.boids = [];
    this.numBoids = 600;
    this.background = '#4499d6';
    this.startTime = Date.now() / 1000;
    this.time = this.startTime;
    this.config = {
      yellow: {
        fillStyle: '#f4bf42',
        radius: 6,
        speed: 2,
        radialSpeed: Math.PI / 60,
        vision: 50,
      },
      salmon: {
        fillStyle: '#fa9466',
        radius: 6,
        speed: 3,
        radialSpeed: Math.PI / 60,
        vision: 80,
      },
      pred: {
        fillStyle: '#f44141',
        radius: 6,
        speed: 3.8,
        radialSpeed: Math.PI / 60,
        vision: 90,
      },
    };
    this.preds = null;
    this.activePred = false;

    this.padding = 8;
  }

  setContext = ctx => (this.ctx = ctx);

  animate = () => this.step();

  createBoids = function(nB) {
    this.numBoids = nB;
    var i;
    for (i = 0; i < (nB || 1); i++) {
      this.boids.push(
        new Boid(this, Math.random() < 0.5 ? 'salmon' : 'yellow')
      );
    }
    this.pred = new Boid(this, 'pred');
  };

  adjustSize = function() {
    var nB = this.numBoids,
      l = this.boids.length,
      even = Math.abs(nB - l) % 2 === 0;
    if (nB < this.boids.length) {
      this.boids.pop();
      if (even) this.boids.pop();
    } else if (nB > this.boids.length) {
      this.boids.push(
        new Boid(this, Math.random() < 0.5 ? 'salmon' : 'yellow')
      );
      if (even)
        this.boids.push(
          new Boid(this, Math.random() < 0.6 ? 'salmon' : 'yellow')
        );
    }
  };

  clear = function() {
    this.boids = [];
  };

  update = function() {
    for (var i = 0; i < this.boids.length; i++) {
      if (this.boids[i].type === 'salmon')
        this.boids[i].update(this.config.salmon);
      else this.boids[i].update(this.config.yellow);
    }
  };

  updatePred = function() {
    this.pred.update(this.config.pred);
  };

  get width() {
    return this.ctx.canvas.width;
  }

  get height() {
    return this.ctx.canvas.height;
  }

  step = function() {
    var ctx = this.ctx;

    var displayWidth = ctx.canvas.parentNode.clientWidth;
    var displayHeight = ctx.canvas.parentNode.clientHeight;

    if (ctx.canvas.width !== displayWidth) ctx.canvas.width = displayWidth;

    if (ctx.canvas.height !== displayHeight) ctx.canvas.height = displayHeight;

    this.adjustSize();

    ctx.fillStyle = this.background;
    ctx.fillRect(0, 0, this.width, this.height);

    for (var i = 0; i < this.boids.length; i++) {
      this.boids[i].step(this);
      this.boids[i].draw(ctx);
    }
    if (this.activePred) {
      this.pred.step(this);
      this.pred.draw(ctx);
    }
  };
}

export default Swarm;
