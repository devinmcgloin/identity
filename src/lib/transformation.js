const flatten = obj => {
  let field = {};

  Object.entries(obj).forEach(([k, v]) => {
    typeof v === 'object' && v !== null && !Array.isArray(v)
      ? Object.assign(field, flatten(v))
      : Object.assign(field, { [k]: v });
  });
  return field;
};

export { flatten };
