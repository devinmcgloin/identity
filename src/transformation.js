const flatten = obj => {
  let field = {};

  Object.entries(obj).forEach(([k, v]) => {
    typeof v === 'object' && v !== null
      ? Object.assign(field, flatten(v))
      : Object.assign(field, { [k]: v });
  });
  return field;
};

export { flatten };
