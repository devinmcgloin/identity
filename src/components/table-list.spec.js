import React from 'react';
import renderer from 'react-test-renderer';
import { flatten } from '../lib/transformation';

import TableList from './table-list';

describe('TableList', () => {
  let data = {
    projects: {
      edges: [
        {
          node: {
            frontmatter: {
              title: 'Sail',
              categories: null,
            },
            fields: {
              slug: '/projects/sail/',
            },
            excerpt: 'Sail is a generative art framework',
          },
        },
        {
          node: {
            frontmatter: {
              title: 'Probabilistic',
              categories: null,
            },
            fields: {
              slug: '/projects/probabilistic/',
            },
            excerpt:
              'Golang implementations of Bloom Filters, Min Sketch Count and Min Hashing.',
          },
        },
        {
          node: {
            frontmatter: {
              title: 'Parking',
              categories: null,
            },
            fields: {
              slug: '/projects/parking/',
            },
            excerpt:
              'Parking is a live art piece in which users claim a parking spot and hopefully make a human connection.',
          },
        },
        {
          node: {
            frontmatter: {
              title: 'clr',
              categories: null,
            },
            fields: {
              slug: '/projects/clr/',
            },
            excerpt:
              'clr is a go library to manage different color spaces, convert between them and compare colors.',
          },
        },
        {
          node: {
            frontmatter: {
              title: 'Fokal',
              categories: null,
            },
            fields: {
              slug: '/projects/fokal/',
            },
            excerpt:
              'Fokal is a simple photography site with the goal of getting good images seen.',
          },
        },
      ],
    },
  };

  let props = {
    columns: [
      { type: 'title', description: 'Title' },
      { type: 'excerpt', description: 'Description' },
    ],
    rows: data.projects.edges.map(e => flatten(e.node)),
    title: 'Projects',
    link: '/projects',
  };

  it('renders correctly with light color', () => {
    const tree = renderer
      .create(<TableList color="light" {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly with dark color', () => {
    const tree = renderer
      .create(<TableList color="dark" {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
