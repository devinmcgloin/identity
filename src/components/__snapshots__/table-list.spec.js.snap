// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableList renders correctly with dark color 1`] = `
<div
  className="pt4 pb5 bg-dark-gray white ma2 br2"
>
  <div
    className="w-100 center mw8"
  >
    <h2
      className="fw7 f3 f2-ns white garamond i"
    >
      Projects
    </h2>
  </div>
  <div
    className="overflow-auto"
  >
    <table
      cellSpacing="0"
      className="f6 w-100 mw8 center"
    >
      <thead>
        <tr>
          <th
            className="fw6 bb white tl pb3 pr3"
          >
            Title
          </th>
          <th
            className="fw6 bb white tl pb3 pr3"
          >
            Description
          </th>
        </tr>
      </thead>
      <tbody
        className="lh-copy"
      >
        <tr>
          <td
            className="pv3 pr3 bb white"
          >
            <a
              className="link dim underline white"
              href="/projects/sail/"
            >
              Sail
            </a>
          </td>
          <td
            className="pv3 pr3 bb white"
          >
            Sail is a generative art framework
          </td>
        </tr>
        <tr>
          <td
            className="pv3 pr3 bb white"
          >
            <a
              className="link dim underline white"
              href="/projects/probabilistic/"
            >
              Probabilistic
            </a>
          </td>
          <td
            className="pv3 pr3 bb white"
          >
            Golang implementations of Bloom Filters, Min Sketch Count and Min Hashing.
          </td>
        </tr>
        <tr>
          <td
            className="pv3 pr3 bb white"
          >
            <a
              className="link dim underline white"
              href="/projects/parking/"
            >
              Parking
            </a>
          </td>
          <td
            className="pv3 pr3 bb white"
          >
            Parking is a live art piece in which users claim a parking spot and hopefully make a human connection.
          </td>
        </tr>
        <tr>
          <td
            className="pv3 pr3 bb white"
          >
            <a
              className="link dim underline white"
              href="/projects/clr/"
            >
              clr
            </a>
          </td>
          <td
            className="pv3 pr3 bb white"
          >
            clr is a go library to manage different color spaces, convert between them and compare colors.
          </td>
        </tr>
        <tr>
          <td
            className="pv3 pr3 bb white"
          >
            <a
              className="link dim underline white"
              href="/projects/fokal/"
            >
              Fokal
            </a>
          </td>
          <td
            className="pv3 pr3 bb white"
          >
            Fokal is a simple photography site with the goal of getting good images seen.
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="w-100 mw8 center tc pt4"
  >
    <a
      className="f6 br2 ba ph3 pv2 mb2 dib near-white hover-dark-gray bg-animate hover-bg-white no-underline index-button"
      href="/projects"
    >
      View More
    </a>
  </div>
</div>
`;

exports[`TableList renders correctly with light color 1`] = `
<div
  className="pt4 pb5 bg-white black ma2 br2"
>
  <div
    className="w-100 center mw8"
  >
    <h2
      className="fw7 f3 f2-ns black garamond i"
    >
      Projects
    </h2>
  </div>
  <div
    className="overflow-auto"
  >
    <table
      cellSpacing="0"
      className="f6 w-100 mw8 center"
    >
      <thead>
        <tr>
          <th
            className="fw6 bb black tl pb3 pr3"
          >
            Title
          </th>
          <th
            className="fw6 bb black tl pb3 pr3"
          >
            Description
          </th>
        </tr>
      </thead>
      <tbody
        className="lh-copy"
      >
        <tr>
          <td
            className="pv3 pr3 bb black"
          >
            <a
              className="link dim underline black"
              href="/projects/sail/"
            >
              Sail
            </a>
          </td>
          <td
            className="pv3 pr3 bb black"
          >
            Sail is a generative art framework
          </td>
        </tr>
        <tr>
          <td
            className="pv3 pr3 bb black"
          >
            <a
              className="link dim underline black"
              href="/projects/probabilistic/"
            >
              Probabilistic
            </a>
          </td>
          <td
            className="pv3 pr3 bb black"
          >
            Golang implementations of Bloom Filters, Min Sketch Count and Min Hashing.
          </td>
        </tr>
        <tr>
          <td
            className="pv3 pr3 bb black"
          >
            <a
              className="link dim underline black"
              href="/projects/parking/"
            >
              Parking
            </a>
          </td>
          <td
            className="pv3 pr3 bb black"
          >
            Parking is a live art piece in which users claim a parking spot and hopefully make a human connection.
          </td>
        </tr>
        <tr>
          <td
            className="pv3 pr3 bb black"
          >
            <a
              className="link dim underline black"
              href="/projects/clr/"
            >
              clr
            </a>
          </td>
          <td
            className="pv3 pr3 bb black"
          >
            clr is a go library to manage different color spaces, convert between them and compare colors.
          </td>
        </tr>
        <tr>
          <td
            className="pv3 pr3 bb black"
          >
            <a
              className="link dim underline black"
              href="/projects/fokal/"
            >
              Fokal
            </a>
          </td>
          <td
            className="pv3 pr3 bb black"
          >
            Fokal is a simple photography site with the goal of getting good images seen.
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    className="w-100 mw8 center tc pt4"
  >
    <a
      className="f6 br2 ba ph3 pv2 mb2 dib dark-gray hover-near-white bg-animate hover-bg-dark-gray no-underline index-button"
      href="/projects"
    >
      View More
    </a>
  </div>
</div>
`;
